File: app/src/main/java/com/udacity/maluleque/meutako/DetailsActivity.java
package com.udacity.maluleque.meutako;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.storage.FirebaseStorage;
import com.squareup.picasso.Picasso;
import com.udacity.maluleque.meutako.model.Transaction;
import com.udacity.maluleque.meutako.utils.NumberUtils;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DetailsActivity extends AppCompatActivity {

    public static final String TRANSACTION = "transaction";
    private static final String TAG = "DetailsActivity";
    private Transaction transaction;
    private FirebaseFirestore db;
    private FirebaseUser user;

    @BindView(R.id.textViewAmount)
    TextView textViewAmount;
    @BindView(R.id.textViewDescription)
    TextView textViewDescription;
    @BindView(R.id.textViewTransactionType)
    TextView textViewType;
    @BindView(R.id.textViewTransationCategory)
    TextView textViewCategory;
    @BindView(R.id.imageView)
    ImageView imageView;
    @BindView(R.id.progressBar)
    ProgressBar progressBar;
    private FirebaseStorage storage;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_details);
        ButterKnife.bind(this);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        db = FirebaseFirestore.getInstance();
        user = FirebaseAuth.getInstance().getCurrentUser();
        storage = FirebaseStorage.getInstance();

        Bundle extras = getIntent().getExtras();
        if (extras.containsKey(TRANSACTION)) {
            transaction = getIntent().getParcelableExtra(TRANSACTION);
        }

        populateData(transaction);
    }

    private void populateData(Transaction transaction) {
        textViewAmount.setText(NumberUtils.getFormattedAmount(transaction.getAmount()));
        textViewCategory.setText(transaction.getCategory());
        textViewDescription.setText(transaction.getDescription());
        textViewType.setText(transaction.getType());

        if (transaction.getImage() != null) {
            if (!transaction.getImage().trim().isEmpty()) {
                progressBar.setVisibility(View.VISIBLE);
                storage.getReference().child(transaction.getImage()).getDownloadUrl().addOnSuccessListener(uri -> {
                    progressBar.setVisibility(View.INVISIBLE);
                    Picasso.get()
                            .load(uri)
                            .centerCrop()
                            .error(R.drawable.no_image)
                            .placeholder(R.drawable.placeholder)
                            .resize(500, 500)
                            .into(imageView);
                }).addOnFailureListener(exception -> {
                    progressBar.setVisibility(View.INVISIBLE);
                    Log.e(TAG, "Error downloading", exception);
                });
            }

        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_details_transaction, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_delete:
                deleteTransaction();
                return true;
            case R.id.action_edit:
                editTransaction();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void editTransaction() {
        Intent intent = new Intent(this, AddTransactionActivity.class);
        intent.putExtra(DetailsActivity.TRANSACTION, transaction);
        startActivity(intent);
    }

    private void deleteTransaction() {
        db.collection("users")
                .document(user.getUid())
                .collection("transactions")
                .document(transaction.getUid())
                .delete()
                .addOnSuccessListener(aVoid -> {
                    Toast.makeText(DetailsActivity.this, "Transaction deleted", Toast.LENGTH_SHORT).show();
                    finish();
                })
                .addOnFailureListener(e -> {
                    Toast.makeText(DetailsActivity.this, "Deletion failed", Toast.LENGTH_SHORT).show();
                    Log.e(TAG, "Error deleting transaction", e);
                });
    }
}

File: app/src/main/java/com/bernardo/maluleque/shibaba/DetailsActivity.kt
package com.bernardo.maluleque.shibaba

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import butterknife.BindView
import butterknife.ButterKnife
import com.bernardo.maluleque.shibaba.model.Transaction
import com.bernardo.maluleque.shibaba.utils.NumberUtils
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.storage.FirebaseStorage
import com.squareup.picasso.Picasso

class DetailsActivity : AppCompatActivity() {
    private var transaction: Transaction? = null
    private var db: FirebaseFirestore? = null
    private var user: FirebaseUser? = null

    @JvmField
    @BindView(R.id.textViewAmount)
    var textViewAmount: TextView? = null

    @JvmField
    @BindView(R.id.textViewDescription)
    var textViewDescription: TextView? = null

    @JvmField
    @BindView(R.id.textViewTransactionType)
    var textViewType: TextView? = null

    @JvmField
    @BindView(R.id.textViewTransationCategory)
    var textViewCategory: TextView? = null

    @JvmField
    @BindView(R.id.imageView)
    var imageView: ImageView? = null

    @JvmField
    @BindView(R.id.progressBar)
    var progressBar: ProgressBar? = null
    private var storage: FirebaseStorage? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_details)
        ButterKnife.bind(this)
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
        db = FirebaseFirestore.getInstance()
        user = FirebaseAuth.getInstance().currentUser
        storage = FirebaseStorage.getInstance()
        val extras = intent.extras
        if (extras!!.containsKey(TRANSACTION)) {
            transaction = intent.getParcelableExtra(TRANSACTION)
        }
        populateData(transaction)
    }

    private fun populateData(transaction: Transaction?) {
        textViewAmount!!.text = NumberUtils.getFormattedAmount(transaction!!.amount)
        textViewCategory!!.text = transaction.category
        textViewDescription!!.text = transaction.description
        textViewType!!.text = transaction.type
        if (transaction.image != null) {
            if (!transaction.image!!.trim { it <= ' ' }.isEmpty()) {
                progressBar!!.visibility = View.VISIBLE
                storage!!.reference.child(transaction.image!!).downloadUrl.addOnSuccessListener { uri: Uri? ->
                    progressBar!!.visibility = View.INVISIBLE
                    Picasso.get()
                            .load(uri)
                            .centerCrop()
                            .error(R.drawable.no_image)
                            .placeholder(R.drawable.placeholder)
                            .resize(500, 500)
                            .into(imageView)
                }.addOnFailureListener { exception: Exception? ->
                    progressBar!!.visibility = View.INVISIBLE
                    Log.e(TAG, "Error downloading", exception)
                }
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_details_transaction, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_delete -> {
                deleteTransaction()
                true
            }
            R.id.action_edit -> {
                editTransaction()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun editTransaction() {
        val intent = Intent(this, AddTransactionActivity::class.java)
        intent.putExtra(TRANSACTION, transaction)
        startActivity(intent)
    }

    private fun deleteTransaction() {
        db!!.collection("users")
                .document(user!!.uid)
                .collection("transactions")
                .document(transaction!!.uid!!)
                .delete()
                .addOnSuccessListener { aVoid: Void? ->
                    Toast.makeText(this@DetailsActivity, "Transaction deleted", Toast.LENGTH_SHORT).show()
                    finish()
                }
                .addOnFailureListener { e: Exception? ->
                    Toast.makeText(this@DetailsActivity, "Deletion failed", Toast.LENGTH_SHORT).show()
                    Log.e(TAG, "Error deleting transaction", e)
                }
    }

    companion object {
        const val TRANSACTION = "transaction"
        private const val TAG = "DetailsActivity"
    }
}